### PR TITLE
Update Connections Overview workbook

### PR DESCRIPTION
**Pull request by andrewki-msft**
Major update to Connections Overview workbook for At-Scale VM Insights.
Various bug fixes such as fixing the inbound direction table entries.
Also showing detailed information for all levels, and not just for
computers. Lastly, add additional filters to only show malicious
connections, links failed, and above average response time. Minor
cosmetic changes to the workbook.

Preview (AtScale)
image

Preview (SingleVM)
image

### CODE REVIEW

**Comment by nasundar** [comment]
Suggestion: Can you merge these to a single dropdown?

image

File: Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook [hunk 1]
```diff
@@ -144,9 +98,7 @@
       "content": {
         "version": "KqlParameterItem/1.0",
         "query": "",
-        "crossComponentResources": [
-          "{Workspace}"
-        ],
+        "crossComponentResources": [],
         "parameters": [
           {
             "id": "306614b4-14f2-4777-a044-8c1121f04d6d",
```
Comment (nasundar) [inline Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook]: If you are not planning on keeping "GridRefresh" please remove it

**Comment by nasundar** [comment]
BUG:

Selecting "Inbound" and a single computer messes the view.

image

File: Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook [hunk 2]
```diff
@@ -413,7 +415,7 @@
...
...d), MaxLinksLive = max(LinksLive), TotalBytesSent = sum(BytesSent), TotalBytesReceived = sum(BytesReceived), AverageResponseTime = 1.0 *  sum(ResponseTimeSum) / sum(Responses) by Computer, ProcessName, RemoteIp, Type = 'Remote Computer', TypeKey = 3, Key = strcat(Computer, '-', ProcessName, '-', tostring(DestinationPort), '-', RemoteIp), ParentKey = strcat(Computer, '-', ProcessName, '-', tostring(DestinationPort)), Id = strcat(Computer, '-', ProcessName, '-', RemoteIp), ComputerName = Computer | join kind=inner     sourcePortData on $left.ParentKey == $right.Key | project-away AverageResponseTime1, Computer1, Key1, LinksFailed1, MaxLinksLive1, ParentKey1, ProcessName1, Responses1, TotalBytesReceived1, TotalBytesSent1, Type1, TypeKey1 | order by Name asc | join kind=leftouter     ipComputerMapping on $left.RemoteIp == $right.Ipv4 | extend Name = iff(RemoteIp == '', 'Unknown',  strcat('🌐 External (', RemoteIp, ')')) | project-away Computer, Ipv4 | order by Name desc | join kind=leftouter     maliciousIpData on $left.Key == $right.MaliciousPortIp | extend MaliciousIpInfo = tostring(MaliciousIpInfo1) | project-away MaliciousIp, Computer, Process, Computer1, MaliciousIpInfo1, Id;\");",
```
Comment (nasundar) [inline Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook]: This is wrong "| where Computer contains ''"

use {ComputerFilter}

File: Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook [hunk 3]
```diff
@@ -441,7 +457,7 @@
...
+            "query": "// {Workspace:label}\r\nprint(\"let sourcePortData = VMConnection | where '{GridRefresh}' == 'run' | where TimeGenerated {TimeRange} | where Direction == '{Direction}' | where '{Direction}' == 'inbound' | where Computer contains '' | summarize Responses = sum(Responses), LinksFailed = sum(LinksFailed), MaxLinksLive = max(LinksLive), TotalBytesSent = sum(BytesSent), TotalBytesReceived = sum(BytesReceived), AverageResponseTime = 1.0 *  sum(ResponseTimeSum) / sum(Responses) by Computer, ProcessName, DestinationIp, DestinationPort, Name = strcat('🔶 ', DestinationPort), Type = 'Remote Port', TypeKey = 4, Key = strcat(Computer, '-', ProcessName, '-', tostring(DestinationPort)), ParentKey = strcat(Computer, '-', ProcessName), ComputerName = Computer | join kind=leftouter maliciousIpData on $left.Key == $right.MaliciousPort | extend MaliciousIpInfo = tostring(MaliciousIpInfo) | project-away Computer1, MaliciousIp, MaliciousPort, MaliciousPortIp, Process; {RemoteIpDataInbound}\");",
```
Comment (nasundar) [inline Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook]: This is wrong -  "| where Computer contains ''"

use {ComputerFilter}

**Comment by nasundar** [comment]
Suggestion: Since the query is split into multiple parts, computerData, processData.... Query construction is taking a lot time. Do some testing to see if the old format of having a single query (May be 2 queries 1 for inbound and 1 for outbound, to simply) is faster.

**Comment by nasundar** [comment]
Since "info" column is hidden malicious info for a connection is also hidden. Will this be brought back?

Suggestion - The info part was added to display details, Computer to display computer level details, process = process level details, port = port level details etc.. May be when you bring this back make changes to display other details as well. 

image

File: Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook [hunk 4]
```diff
@@ -525,7 +597,7 @@
             "name": "FinalQuery",
             "type": 1,
             "isRequired": false,
-            "query": "print(\r\niff({Hierarchy} == 0, \"{Computer_Process_IP_Port}\", \r\niff({Hierarchy} == 1, \"{Computer_Process_IP}\",\r\niff({Hierarchy} == 2, \"{Computer_Process}\",\r\niff({Hierarchy} == 3, \"{Computer}\", \"{Computer}\")))));",
+            "query": "print(strcat(\r\niff({Hierarchy} == 0, \"{Computer_Process_IP_Port}\", \r\niff({Hierarchy} == 1, \"{Computer_Process_IP}\",\r\niff({Hierarchy} == 2, \"{Computer_Process}\",\r\niff({Hierarchy} == 3, \"{Computer}\", \"{Computer}\")))),' | where MaliciousConnectionsCount >= {OnlyMaliciousConnections}',' | where LinksFailed >= {OnlyLinksFailed}',' | where AverageResponseTime >= {HighResponseTime}'));",
             "crossComponentResources": [
```
Comment (nasundar) [inline Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook]: For HighResponseTime, once we have tool tips add info saying > 1sec.

File: Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook [hunk 5]
```diff
@@ -733,37 +815,101 @@
...
-            "id": "4c14e89e-dbd3-421b-a041-c8fc87ce6e43",
+            "id": "7e4c3d29-2be6-4288-903b-95502ddab577",
             "version": "KqlParameterItem/1.0",
-            "name": "SourceComputer",
+            "name": "ComputerName",
             "type": 1,
-            "isRequired": true,
-            "query": "let row = dynamic({ConnectionGrid});\r\nlet computer = tostring(row.Key);\r\nrange i from 1 to 1 step 1\r\n| project computer",
+            "isRequired": false,
+            "query": "// {Workspace:label}\r\nlet row = dynamic({ConnectionGrid});\r\nlet computerName = row.ComputerName;\r\nprint computerName",
```
Comment (nasundar) [inline Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook]: I see this comment "// {Workspace:label}" in almost all queries. Do you need that?

**Comment by nasundar** [comment]
I see you have checked in "Connection Overview - Single VM" even that has the same bug - {ComputerFilter}, please fix it.

**Comment by nasundar** [comment]
For Single VM, We can technically remove workspace restriction and do Resource centric queries. I think I had asked this in the previous review as well. Do you have plans of using Resource Centric for Single VM?

**Comment by andrewki-msft** [comment]
> BUG:
> Selecting "Inbound" and a single computer messes the view.

Thanks for catching this, will fix this in both AtScale and SingleVM

**Comment by andrewki-msft** [comment]
Also will remove `where Computer contains ''` from Single VM Connection

**Comment by nasundar** [comment]
Can you reword this?
Data might not be present for the selected time range and/or for selected filters - Malicious Connections, Failed Connections etc,, So you have make it clear for user that he needs to change filters.

image

File: Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook (hunk 3 above)
Comment (andrewki-msft) [reply Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook]: Done

File: Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook (hunk 2 above)
Comment (andrewki-msft) [reply Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook]: Done

File: Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook (hunk 5 above)
Comment (andrewki-msft) [reply Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook]: Removed, they were originally added as a _safe_ workaround to guarantee the template would initialize properly under certain conditions but looks like that has been fixed

File: Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook (hunk 4 above)
Comment (andrewki-msft) [reply Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook]: The new dropdown looks like this, so should be self-explanatory for the time being
image

File: Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook (hunk 1 above)
Comment (andrewki-msft) [reply Workbooks/Virtual Machines - Network Dependencies/Connections Overview/Connections Overview.workbook]: Done

**Comment by andrewki-msft** [comment]
> Suggestion: Since the query is split into multiple parts, computerData, processData.... Query construction is taking a lot time. Do some testing to see if the old format of having a single query (May be 2 queries 1 for inbound and 1 for outbound, to simply) is faster.

I won't merge the queries together since a lot of effort has been put into separating them out for easier editing. I'll bring this up with the workbooks team in our next sync-up to see what ideas we can do to better support _Query Snippets_.

**Comment by andrewki-msft** [comment]
> Since "info" column is hidden malicious info for a connection is also hidden. Will this be brought back?
> Suggestion - The info part was added to display details, Computer to display computer level details, process = process level details, port = port level details etc.. May be when you bring this back make changes to display other details as well.

The PM requested the info column to be removed (as opposed to the Performance workbook where it makes sense there), your suggestion is reasonable but you'll need to pass this up with the PM to explicitly state what info you want in there that's not shown in the table.

**Comment by nasundar** [comment]
Single VM Connections Overview, selecting a row does not work.

image

**Comment by nasundar** [comment]
Selecting "HighResponseTime" filter has dangling nodes.

image

**Comment by nasundar** [comment]
Sure, I will check with PMs.

**Pull request merged**